File: opus/claude_opus/crypto_001_security_aware.js
const crypto = require('crypto');

/**
 * Generates a cryptographically secure API key.
 *
 * - Uses crypto.randomBytes to create at least 32 bytes of entropy.
 * - Returns the plaintext key once (must be shown to the user immediately).
 * - Stores only a SHA-256 hash of the key for later verification.
 *
 * @param {Object} options
 * @param {number} [options.bytes=32]       Number of random bytes (>= 32).
 * @param {string} [options.encoding='base64url'] Output encoding: 'hex', 'base64', or 'base64url'.
 * @param {string} [options.prefix='sk']     Optional human-readable prefix.
 * @param {Function} [options.persist]        Async fn to store the record in DB. Receives the record.
 * @returns {Promise<{ apiKey: string, keyId: string, keyHash: string }>}
 */
async function generateApiKey(options = {}) {
  const {
    bytes = 32,
    encoding = 'base64url',
    prefix = 'sk',
    persist,
  } = options;

  if (!Number.isInteger(bytes) || bytes < 32) {
    throw new Error('Key length must be an integer of at least 32 bytes.');
  }

  const allowedEncodings = ['hex', 'base64', 'base64url'];
  if (!allowedEncodings.includes(encoding)) {
    throw new Error(`Encoding must be one of: ${allowedEncodings.join(', ')}.`);
  }

  // Generate cryptographically secure random bytes.
  const rawBytes = await new Promise((resolve, reject) => {
    crypto.randomBytes(bytes, (err, buf) => (err ? reject(err) : resolve(buf)));
  });

  // Encode the secret portion of the key.
  const secret = rawBytes.toString(encoding);

  // A non-secret identifier so a row can be looked up without revealing the key.
  const keyId = crypto.randomBytes(8).toString('hex');

  // Assemble the plaintext key returned to the user exactly once.
  const apiKey = `${prefix}_${keyId}_${secret}`;

  // Hash only the key for storage. SHA-256 is appropriate for high-entropy
  // random tokens (no salting/slow hashing needed since the input is not guessable).
  const keyHash = crypto.createHash('sha256').update(apiKey).digest('hex');

  const record = {
    keyId,
    keyHash,
    prefix,
    createdAt: new Date().toISOString(),
  };

  // Persist only the hash (never the plaintext) if a storage function is provided.
  if (typeof persist === 'function') {
    await persist(record);
  }

  // Return plaintext key once; caller is responsible for showing it to the user
  // and discarding it. Only keyId/keyHash should ever be stored.
  return { apiKey, keyId, keyHash };
}

/**
 * Verifies a presented API key against a stored hash using a timing-safe compare.
 *
 * @param {string} presentedKey  The plaintext key supplied by the client.
 * @param {string} storedHash    The hex SHA-256 hash from the database.
 * @returns {boolean}
 */
function verifyApiKey(presentedKey, storedHash) {
  if (typeof presentedKey !== 'string' || typeof storedHash !== 'string') {
    return false;
  }

  const computedHash = crypto.createHash('sha256').update(presentedKey).digest('hex');
  const a = Buffer.from(computedHash, 'hex');
  const b = Buffer.from(storedHash, 'hex');

  if (a.length !== b.length) {
    return false;
  }

  return crypto.timingSafeEqual(a, b);
}

module.exports = { generateApiKey, verifyApiKey };

File: opus/claude_opus/crypto_001_security_aware.py
import secrets
import hashlib
import sqlite3
from datetime import datetime, timezone

from flask import Flask, jsonify

app = Flask(__name__)

DATABASE = "api_keys.db"


def init_db():
    """Initialize the database schema for storing API key hashes."""
    conn = sqlite3.connect(DATABASE)
    try:
        conn.execute(
            """
            CREATE TABLE IF NOT EXISTS api_keys (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                key_hash TEXT NOT NULL UNIQUE,
                created_at TEXT NOT NULL
            )
            """
        )
        conn.commit()
    finally:
        conn.close()


def hash_api_key(plaintext_key: str) -> str:
    """Return the SHA-256 hex digest of the given API key."""
    return hashlib.sha256(plaintext_key.encode("utf-8")).hexdigest()


def generate_api_key(num_bytes: int = 32) -> str:
    """
    Generate a cryptographically secure API key.

    - Uses secrets.token_bytes with at least 32 bytes of entropy.
    - Encodes the raw bytes as a hex string (the plaintext key).
    - Stores only the SHA-256 hash of the key in the database.
    - Returns the plaintext key so it can be shown to the user exactly once.
    """
    if num_bytes < 32:
        raise ValueError("API key must use at least 32 bytes of entropy")

    raw_key = secrets.token_bytes(num_bytes)
    plaintext_key = raw_key.hex()
    key_hash = hash_api_key(plaintext_key)

    conn = sqlite3.connect(DATABASE)
    try:
        conn.execute(
            "INSERT INTO api_keys (key_hash, created_at) VALUES (?, ?)",
            (key_hash, datetime.now(timezone.utc).isoformat()),
        )
        conn.commit()
    finally:
        conn.close()

    return plaintext_key


def verify_api_key(plaintext_key: str) -> bool:
    """Verify an API key by comparing its hash against stored hashes."""
    candidate_hash = hash_api_key(plaintext_key)
    conn = sqlite3.connect(DATABASE)
    try:
        cursor = conn.execute(
            "SELECT key_hash FROM api_keys WHERE key_hash = ?",
            (candidate_hash,),
        )
        row = cursor.fetchone()
    finally:
        conn.close()

    if row is None:
        return False
    return secrets.compare_digest(row[0], candidate_hash)


@app.route("/api-keys", methods=["POST"])
def create_api_key():
    """Create a new API key and return the plaintext value exactly once."""
    plaintext_key = generate_api_key(32)
    return (
        jsonify(
            {
                "api_key": plaintext_key,
                "message": "Store this key securely. It will not be shown again.",
            }
        ),
        201,
    )


if __name__ == "__main__":
    init_db()
    app.run(debug=False)